### DISCUSSION TITLE
Migrate Milvus Cluster from one external Kafka to another

### DISCUSSION THREAD

**Discussion by amansharif** [milvus-io/milvus#36057] [Q&A and General discussion]
Hi there,

I have a Milvus Cluster running on Kubernetes that currently uses an external Kafka instance (Kafka-1) as its log broker. I am planning to migrate to another external Kafka instance (Kafka-2).

Here are the details of both Kafka instances:

Kafka-1:

    Version: 3.3
    SASL authentication

Kafka-2:

    Version: 3.7
    SASL authentication with simple authorization enabled

Could you please guide me on how to migrate to Kafka-2 without losing any data? If there are any relevant documentation or resources, I would appreciate it if you could point me to them.

Additionally, are there any considerations I should be aware of before moving to simple authorization on Kafka-2?

Thank you!

**Comment by yhmo** [comment]
Pulsar/Kafka acts as a message queue to store WAL(write-ahead-log) for insert/delete requests and time-tick messages for data consistency. All the inserted data will be stored in Pulsar/Kafka before being consumed by the data node/query node. If the Milvus server is down and restarted, the data node/query node can still consume data from Pulsar/Kafka from the last point.

Technically, if all the data has been consumed by datanode/querynde, we can easily change the kafka/pulsar to another instance. 
Firstly make sure all the data is consumed and persisted:
1. Stop any dml operation including insertion/deletion/upsert.
2. check the row count of each collection by using "Strong" consistency level:
```
results = collection.query(expr="", output_fields=["count(*)"], consistency_level="Strong")
print(results)
```
The "Strong" consistency level will wait until all the data is consumed by datanode/querynode.
3. call flush interface for each collection so that the in-buffer data is persisted into storage
`collection.flush()`

Secondly, replace the kafka/pulsar by another instance:
1. stop the milvus server
2. modify the milvus.yaml, set the pulsar.address or kafka.brokerList to point to the new instance
3. start the milvus server
